File: data/two-pointers/move-zeroes.cpp
#include <iostream>
#include <vector>
using namespace std;

void moveZeroes(vector<int>& nums) {
    int slow = 0;  // Pointer for placing non-zero elements

    // Iterate through the array
    for (int fast = 0; fast < nums.size(); fast++) {
        if (nums[fast] != 0) {
            swap(nums[slow], nums[fast]);  // Swap non-zero element to correct position
            slow++;  // Move slow pointer forward
        }
    }
}

// Example Usage
int main() {
    vector<int> nums = {0, 1, 0, 3, 12};
    moveZeroes(nums);
    
    // Print result
    for (int num : nums) {
        cout << num << " ";
    }
    return 0;
}

File: data/two-pointers/move-zeroes.py
def moveZeroes(nums):
    """
    Moves all zeros in nums to the end while maintaining the relative order of non-zero elements.

    :param nums: List[int] - Input array
    :return: None - Modifies nums in-place
    """
    slow = 0  # Pointer for placing non-zero elements

    # Move non-zero elements forward
    for fast in range(len(nums)):
        if nums[fast] != 0:
            nums[slow], nums[fast] = nums[fast], nums[slow]  # Swap
            slow += 1  # Move slow pointer forward

# Example Usage
nums = [0, 1, 0, 3, 12]
moveZeroes(nums)
print(nums)  # Output: [1, 3, 12, 0, 0]
